src/Optimizer.ts

```diff
@@ -0,0 +1,102 @@
+import { Report, ReportElement } from './Models/Report';
+import { parse } from '@asyncapi/parser';
+import { RemoveComponents } from './Optimizers/RemoveComponents';
+import { MoveToComponents } from './Optimizers/MoveToComponents';
+import { ReuseComponents } from './Optimizers/ReuseComponents';
+import { Options } from './Models/Options';
+import YAML from 'yaml';
+import merge from 'merge-deep';
+import * as _ from 'lodash';
+import { ComponentProvider } from './ComponentProvider';
+/**
+ * this class is the starting point of the library.
+ * user will only interact with this class. here we generate different kind of reports using optimizers, apply changes and return the results to the user.
+ *
+ * @public
+ */
+export class Optimizer {
+  componentProvider!: ComponentProvider;
+  reuseComponentsReport: ReportElement[] = [];
+  removeComponentsReport: ReportElement[] = [];
+  moveToComponentsReport: ReportElement[] = [];
+  outputObject = {};
+  constructor(private YAMLorJSON: any) {
+
+  }
+
+  /**
+   * @returns a {@link Report} object containing all of the optimizations that the library can do.
+   * @public
+   */
+  getReport = async (): Promise<Report> => {
+    if (!this.componentProvider) {
+      const parsedDocument = await parse(this.YAMLorJSON);
+      this.componentProvider = new ComponentProvider(parsedDocument);
+    }
+    const reuseComponents = new ReuseComponents(this.componentProvider);
+    this.reuseComponentsReport = reuseComponents.getReport();
+    const removeComponents = new RemoveComponents(this.componentProvider);
+    this.removeComponentsReport = removeComponents.getReport();
+    const moveToComponents = new MoveToComponents(this.componentProvider);
+    this.moveToComponentsReport = moveToComponents.getReport();
+
+    return {
+      reuseComponents: this.reuseComponentsReport,
+      removeComponents: this.removeComponentsReport,
+      moveToComponents: this.moveToComponentsReport
+    };
+  }
+  sortFunction = (a: ReportElement, b: ReportElement): number => b.path.length - a.path.length;
+  getOptimizedDocument = (options: Options): string => {
+    const defaultOptions = {
+      rules: {
+        reuseComponents: true,
+        removeComponents: true,
+        moveToComponents: true
+      }
+    };
+    options = merge(defaultOptions, options);
+    this.outputObject = YAML.parse(this.YAMLorJSON);
```

Comment:
I did not investigate deeper but looks to me like library supports YAML only, as when I pass JSON as string or object, it fails.

also readme says 
```
#### new Optimizer(document)

`document` is a mandatory object which is the parsed AsyncAPI specification from `@asyncapi/parser`:
```

but I believe it is not true

Reply:
I am planning to add support for JSON and string JSONs. but this PR has become too complicated to add new things to it. so I am opening an issue to support them in the near future (after this PR.).

Reply:
@KhudaDad414 good decision, then just fix issue with readme and traits and let us party merge 😄